Windows cmake 3.21.4 build fails

Hello,

When compiling the Anari SDK on Windows 11 Visual Studio 2019 I get this build error:

`8>Generating Code...
8>Done building project "anari_library_example.vcxproj" -- FAILED.
12>------ Build started: Project: anariTutorialDirectLink, Configuration: Release x64 ------
12>Building Custom Rule C:/Users/Cameron/Downloads/ANARI-SDK/examples/direct_linking/CMakeLists.txt
12>anariTutorialDirectLink.c
12>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_library_example.lib'
12>Done building project "anariTutorialDirectLink.vcxproj" -- FAILED.
13>------ Build started: Project: ALL_BUILD, Configuration: Release x64 ------
13>Building Custom Rule C:/Users/Cameron/Downloads/ANARI-SDK/CMakeLists.txt
========== Build: 8 succeeded, 5 failed, 0 up-to-date, 0 skipped ==========
`
I attempt to continue the build:

`1>C:\Users\Cameron\Downloads\ANARI-SDK\examples\example_device\scene\bvh/SweepSAHBuilder.h(294,3): error C2059: syntax error: '{'
1>C:\Users\Cameron\Downloads\ANARI-SDK\examples\example_device\scene\bvh/SweepSAHBuilder.h(294,3): error C2143: syntax error: missing ';' before '{'
1>C:\Users\Cameron\Downloads\ANARI-SDK\examples\example_device\scene\bvh/SweepSAHBuilder.h(301,31): warning C4267: '=': conversion from 'size_t' to 'uint32_t', possible loss of data
1>C:\Users\Cameron\Downloads\ANARI-SDK\examples\example_device\scene\volume\../SceneObject.h(84,9): error C3001: 'declare': expected an OpenMP directive name
1>C:\Users\Cameron\Downloads\ANARI-SDK\examples\example_device\scene\volume\../SceneObject.h(88,9): error C3036: 'bbox_extend': invalid operator token in OpenMP 'reduction' clause
1>C:\Users\Cameron\Downloads\ANARI-SDK\examples\example_device\scene\volume\../SceneObject.h(88,1): error C3016: 'i': index variable in OpenMP 'for' statement must have signed integral type
1>Generating Code...
1>Done building project "anari_library_example.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 3 up-to-date, 0 skipped ==========`

Am I missing prerequisites? I hope this is helpful.

Kind regards,

Cameron

Hi Jefferson,

Thanks very much, commenting that out seems to have improved the compilation on Windows.

This is the remaining result, I have removed the Warnings.

`6>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\simple\anariTutorial.c(13,10): fatal error C1083: Cannot open include file: 'alloca.h': No such file or directory
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
10>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_utilities.lib'
11>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_utilities.lib'
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\renderer\Renderer.cpp(20,51): error C7555: use of designated initializers requires at least '/std:c++20'
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\scene\bvh/BVH.h(66,33): error C2338: BVH::Node is too big!
12>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_library_example.lib'
`
Many thanks and kind regards,

Cameron

Hi Jefferson,

Redownloaded main and compiled, it's looking better.

`6>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\simple\anariTutorial.c(13,10): fatal error C1083: Cannot open include file: 'alloca.h': No such file or directory
11>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_utilities.lib'
10>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_utilities.lib'
8>C:\Users\Cameron\Downloads\ANARI-SDK-main\examples\example_device\renderer\Renderer.cpp(20,51): error C7555: use of designated initializers requires at least '/std:c++20'
12>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_library_example.lib'

...
8>Done building project "anari_library_example.vcxproj" -- FAILED.
12>------ Build started: Project: anariTutorialDirectLink, Configuration: Release x64 ------
12>Building Custom Rule C:/Users/Cameron/Downloads/ANARI-SDK-main/examples/direct_linking/CMakeLists.txt
12>anariTutorialDirectLink.c
12>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_library_example.lib'
12>Done building project "anariTutorialDirectLink.vcxproj" -- FAILED.
13>------ Build started: Project: ALL_BUILD, Configuration: Release x64 ------T
13>Building Custom Rule C:/Users/Cameron/Downloads/ANARI-SDK-main/CMakeLists.txt
========== Build: 8 succeeded, 5 failed, 0 up-to-date, 0 skipped ==========`

Thank you for the quick feedback!

Kind regards,,

Cameron

Its better.

13>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_utilities.lib'
13>Done building project "anari_library_example.vcxproj" -- FAILED.
17>------ Rebuild All started: Project: anariTutorialDirectLink, Configuration: Release x64 ------
17>Building Custom Rule C:/Users/Cameron/Downloads/ANARI-SDK-main/examples/direct_linking/CMakeLists.txt
17>anariTutorialDirectLink.c
17>LINK : fatal error LNK1181: cannot open input file '..\..\Release\anari_library_example.lib'
17>Done building project "anariTutorialDirectLink.vcxproj" -- FAILED.
18>------ Rebuild All started: Project: ALL_BUILD, Configuration: Release x64 ------
18>Building Custom Rule C:/Users/Cameron/Downloads/ANARI-SDK-main/CMakeLists.txt
19>------ Skipped Rebuild All: Project: INSTALL, Configuration: Release x64 ------
19>Project not selected to build for this solution configuration 
========== Rebuild All: 9 succeeded, 4 failed, 6 skipped ==========

-I appreciate you're tweaking all this from a Linux based codebase.

Many thanks,

Cameron

--- a/examples/example_device/CMakeLists.txt
+++ b/examples/example_device/CMakeLists.txt
@@ -89,7 +89,7 @@ endif()
 target_compile_definitions(${PROJECT_NAME}
 PRIVATE
   -DDEVICE_VERSION=${REFERENCE_DEVICE_VERSION}
-  "anari_EXPORTS"
+  "example_device_EXPORTS"
 )
 
 install(TARGETS ${PROJECT_NAME}

--- a/examples/example_device/CMakeLists.txt
+++ b/examples/example_device/CMakeLists.txt
@@ -89,6 +89,7 @@ endif()
 target_compile_definitions(${PROJECT_NAME}
 PRIVATE
   -DDEVICE_VERSION=${REFERENCE_DEVICE_VERSION}
+  "anari_EXPORTS"
 )
 
 install(TARGETS ${PROJECT_NAME}

--- a/examples/example_device/ExampleDevice.cpp
+++ b/examples/example_device/ExampleDevice.cpp
@@ -4,6 +4,7 @@
 #include "ExampleDevice.h"
 
 #include "anari/detail/Library.h"
+#include "include/anari/devices/example.h"
 
 #include "Array.h"
 #include "frame/Frame.h"

--- a/examples/example_device/ExampleDevice.cpp
+++ b/examples/example_device/ExampleDevice.cpp
@@ -532,24 +532,27 @@ void ExampleDevice::flushCommitBuffer()
 
 static char deviceName[] = "example";
 
-ANARI_DEFINE_LIBRARY_NEW_DEVICE(example, subtype)
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARI_DEFINE_LIBRARY_NEW_DEVICE(
+    example, subtype)
 {
   if (subtype == std::string("default") || subtype == std::string("example"))
     return (ANARIDevice) new anari::example_device::ExampleDevice();
   return nullptr;
 }
 
-ANARI_DEFINE_LIBRARY_INIT(example)
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARI_DEFINE_LIBRARY_INIT(example)
 {
   printf("...loaded example library!\n");
 }
 
-ANARI_DEFINE_LIBRARY_GET_DEVICE_SUBTYPES(example, libdata)
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARI_DEFINE_LIBRARY_GET_DEVICE_SUBTYPES(
+    example, libdata)
 {
   static const char *devices[] = {deviceName, nullptr};
   return devices;
 }
-ANARI_DEFINE_LIBRARY_GET_OBJECT_SUBTYPES(
+
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARI_DEFINE_LIBRARY_GET_OBJECT_SUBTYPES(
     example, libdata, deviceSubtype, objectType)
 {
   if (objectType == ANARI_RENDERER) {
@@ -563,14 +566,17 @@ ANARI_DEFINE_LIBRARY_GET_OBJECT_SUBTYPES(
   }
   return nullptr;
 }
-ANARI_DEFINE_LIBRARY_GET_OBJECT_PARAMETERS(
+
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARI_DEFINE_LIBRARY_GET_OBJECT_PARAMETERS(
     example, libdata, deviceSubtype, objectSubtype, objectType)
 {
   if (objectType == ANARI_RENDERER)
     return anari::example_device::Renderer::g_parameters;
   return nullptr;
 }
-ANARI_DEFINE_LIBRARY_GET_PARAMETER_PROPERTY(example,
+
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARI_DEFINE_LIBRARY_GET_PARAMETER_PROPERTY(
+    example,
     libdata,
     deviceSubtype,
     objectSubtype,
@@ -606,7 +612,7 @@ ANARI_DEFINE_LIBRARY_GET_PARAMETER_PROPERTY(example,
   return nullptr;
 }
 
-extern "C" ANARIDevice anariNewExampleDevice()
+extern "C" EXAMPLE_DEVICE_INTERFACE ANARIDevice anariNewExampleDevice()
 {
   return (ANARIDevice) new anari::example_device::ExampleDevice();
 }

--- a/examples/example_device/ExampleDevice.h
+++ b/examples/example_device/ExampleDevice.h
@@ -10,14 +10,16 @@
 // std
 #include <vector>
 
+#include "interface.h"
+
 namespace anari {
 namespace example_device {
 
 struct Object;
 
-struct ExampleDevice : public Device,
-                         public RefCounted,
-                         public ParameterizedObject
+struct EXAMPLE_DEVICE_INTERFACE ExampleDevice : public Device,
+                                                public RefCounted,
+                                                public ParameterizedObject
 {
   /////////////////////////////////////////////////////////////////////////////
   // Main interface to accepting API calls

--- a/examples/example_device/include/anari/devices/example.h
+++ b/examples/example_device/include/anari/devices/example.h
@@ -5,11 +5,13 @@
 
 #include "anari/anari.h"
 
+#include "../../../interface.h"
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-ANARI_INTERFACE ANARIDevice anariNewExampleDevice();
+EXAMPLE_DEVICE_INTERFACE ANARIDevice anariNewExampleDevice();
 
 #ifdef __cplusplus
 } // extern "C"

--- a/examples/example_device/include/anari/devices/example.h
+++ b/examples/example_device/include/anari/devices/example.h
@@ -9,7 +9,7 @@
 extern "C" {
 #endif
 
-ANARIDevice anariNewExampleDevice();
+ANARI_INTERFACE ANARIDevice anariNewExampleDevice();
 
 #ifdef __cplusplus
 } // extern "C"

--- a/examples/example_device/interface.h
+++ b/examples/example_device/interface.h
@@ -1,0 +1,20 @@
+// Copyright 2021 The Khronos Group
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#ifdef _WIN32
+#ifdef EXAMPLE_DEVICE_STATIC_DEFINE
+#define EXAMPLE_DEVICE_INTERFACE
+#else
+#ifdef example_device_EXPORTS
+#define EXAMPLE_DEVICE_INTERFACE __declspec(dllexport)
+#else
+#define EXAMPLE_DEVICE_INTERFACE __declspec(dllimport)
+#endif
+#endif
+#elif defined __GNUC__
+#define EXAMPLE_DEVICE_INTERFACE __attribute__((__visibility__("default")))
+#else
+#define EXAMPLE_DEVICE_INTERFACE
+#endif

--- a/examples/example_device/renderer/Renderer.cpp
+++ b/examples/example_device/renderer/Renderer.cpp
@@ -17,7 +17,7 @@ ANARIParameter Renderer::g_parameters[] = {
     {NULL, ANARI_UNKNOWN}};
 
 ParameterInfo::Def::Def(int i) : i(i) {}
-ParameterInfo::Def::Def(vec4 v) : v(v) {};
+ParameterInfo::Def::Def(vec4 v) : v(v) {}
 
 ParameterInfo Renderer::g_parameterinfos[] = {
     {"background color and alpha (RGBA)", false, {vec4(1.f)}},

--- a/examples/example_device/renderer/Renderer.cpp
+++ b/examples/example_device/renderer/Renderer.cpp
@@ -16,8 +16,11 @@ ANARIParameter Renderer::g_parameters[] = {
     {"pixelSamples", ANARI_INT32},
     {NULL, ANARI_UNKNOWN}};
 
+ParameterInfo::Def::Def(int i) : i(i) {}
+ParameterInfo::Def::Def(vec4 v) : v(v) {};
+
 ParameterInfo Renderer::g_parameterinfos[] = {
-    {"background color and alpha (RGBA)", false, {.v = vec4(1.f)}},
+    {"background color and alpha (RGBA)", false, {vec4(1.f)}},
     {"samples per pixel", false, {1}}};
 
 FactoryMapPtr<Renderer> Renderer::g_renderers;

--- a/examples/example_device/renderer/Renderer.h
+++ b/examples/example_device/renderer/Renderer.h
@@ -14,10 +14,12 @@ typedef struct
 {
   const char *desc;
   bool req;
-  union
+  union Def
   {
     int i;
     vec4 v;
+    Def(int i);
+    Def(vec4 v);
   } def;
 } ParameterInfo;
 

--- a/examples/simple/anariTutorial.c
+++ b/examples/simple/anariTutorial.c
@@ -10,7 +10,11 @@
 //   raster renderer, there is no improvement to the initial rendering, so
 //   both images will appear the same.
 
+#ifdef _WIN32
+#include <malloc.h>
+#else
 #include <alloca.h>
+#endif
 #include <errno.h>
 #include <stdint.h>
 #include <stdio.h>

--- a/examples/viewer/OrbitManipulator.cpp
+++ b/examples/viewer/OrbitManipulator.cpp
@@ -9,7 +9,7 @@ namespace manipulators {
 
 static float degreesToRadians(float degrees)
 {
-  return degrees * M_PI / 180.f;
+  return degrees * glm::pi<float>() / 180.f;
 }
 
 static glm::vec3 azelToDirection(float az, float el, OrbitAxis axis)

--- a/libs/anari/API.cpp
+++ b/libs/anari/API.cpp
@@ -522,7 +522,7 @@ extern "C" void anariDiscardFrame(ANARIDevice d, ANARIFrame f) ANARI_CATCH_BEGIN
 }
 ANARI_CATCH_END_NORETURN()
 
-ANARI_INTERFACE ANARIObject anariNewObject(
+ANARIObject anariNewObject(
     ANARIDevice d, const char *objectType, const char *type) ANARI_CATCH_BEGIN
 {
   return deviceRef(d).newObject(objectType, type);

--- a/libs/anari/Device.cpp
+++ b/libs/anari/Device.cpp
@@ -10,13 +10,12 @@ namespace anari {
 
 // extension functions ////////////////////////////////////////////////////////
 
-ANARI_INTERFACE ANARIObject Device::newObject(
-    const char *objectType, const char *type)
+ANARIObject Device::newObject(const char *objectType, const char *type)
 {
   return nullptr;
 }
 
-ANARI_INTERFACE void (*Device::getProcAddress(const char *name))(void)
+void (*Device::getProcAddress(const char *name))(void)
 {
   return nullptr;
 }

--- a/libs/anari/include/anari/detail/Device.h
+++ b/libs/anari/include/anari/detail/Device.h
@@ -175,7 +175,7 @@ struct Device
       void *defaultStatusCBUserPtr = nullptr);
 
  protected:
-  ANARIDevice this_device() const;
+  ANARI_INTERFACE ANARIDevice this_device() const;
 
   ANARI_INTERFACE ANARIStatusCallback defaultStatusCallback() const;
   ANARI_INTERFACE void *defaultStatusCallbackUserPtr() const;

--- a/libs/anari/include/anari/detail/Device.h
+++ b/libs/anari/include/anari/detail/Device.h
@@ -29,7 +29,7 @@ inline B *allocate_object(Args... args)
   return new T(args...);
 }
 
-struct Device
+struct ANARI_INTERFACE Device
 {
   /////////////////////////////////////////////////////////////////////////////
   // Main virtual interface to accepting API calls
@@ -37,29 +37,29 @@ struct Device
 
   // Device API ///////////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual int deviceImplements(const char *extension) = 0;
+  virtual int deviceImplements(const char *extension) = 0;
 
-  ANARI_INTERFACE virtual void deviceSetParameter(
+  virtual void deviceSetParameter(
       const char *id, ANARIDataType type, const void *mem) = 0;
 
-  ANARI_INTERFACE virtual void deviceUnsetParameter(const char *id) = 0;
+  virtual void deviceUnsetParameter(const char *id) = 0;
 
-  ANARI_INTERFACE virtual void deviceCommit() = 0;
+  virtual void deviceCommit() = 0;
 
-  ANARI_INTERFACE virtual void deviceRetain() = 0;
+  virtual void deviceRetain() = 0;
 
-  ANARI_INTERFACE virtual void deviceRelease() = 0;
+  virtual void deviceRelease() = 0;
 
   // Data Arrays //////////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIArray1D newArray1D(void *appMemory,
+  virtual ANARIArray1D newArray1D(void *appMemory,
       ANARIMemoryDeleter deleter,
       void *userdata,
       ANARIDataType,
       uint64_t numItems1,
       uint64_t byteStride1) = 0;
 
-  ANARI_INTERFACE virtual ANARIArray2D newArray2D(void *appMemory,
+  virtual ANARIArray2D newArray2D(void *appMemory,
       ANARIMemoryDeleter deleter,
       void *userdata,
       ANARIDataType,
@@ -68,7 +68,7 @@ struct Device
       uint64_t byteStride1,
       uint64_t byteStride2) = 0;
 
-  ANARI_INTERFACE virtual ANARIArray3D newArray3D(void *appMemory,
+  virtual ANARIArray3D newArray3D(void *appMemory,
       ANARIMemoryDeleter deleter,
       void *userdata,
       ANARIDataType,
@@ -79,57 +79,54 @@ struct Device
       uint64_t byteStride2,
       uint64_t byteStride3) = 0;
 
-  ANARI_INTERFACE virtual void *mapArray(ANARIArray) = 0;
-  ANARI_INTERFACE virtual void unmapArray(ANARIArray) = 0;
+  virtual void *mapArray(ANARIArray) = 0;
+  virtual void unmapArray(ANARIArray) = 0;
 
   // Renderable Objects ///////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARILight newLight(const char *type) = 0;
+  virtual ANARILight newLight(const char *type) = 0;
 
-  ANARI_INTERFACE virtual ANARICamera newCamera(const char *type) = 0;
+  virtual ANARICamera newCamera(const char *type) = 0;
 
-  ANARI_INTERFACE virtual ANARIGeometry newGeometry(const char *type) = 0;
-  ANARI_INTERFACE virtual ANARISpatialField newSpatialField(
-      const char *type) = 0;
+  virtual ANARIGeometry newGeometry(const char *type) = 0;
+  virtual ANARISpatialField newSpatialField(const char *type) = 0;
 
-  ANARI_INTERFACE virtual ANARISurface newSurface() = 0;
-  ANARI_INTERFACE virtual ANARIVolume newVolume(const char *type) = 0;
+  virtual ANARISurface newSurface() = 0;
+  virtual ANARIVolume newVolume(const char *type) = 0;
 
   // Surface Meta-Data ////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIMaterial newMaterial(
-      const char *material_type) = 0;
+  virtual ANARIMaterial newMaterial(const char *material_type) = 0;
 
-  ANARI_INTERFACE virtual ANARISampler newSampler(const char *type) = 0;
+  virtual ANARISampler newSampler(const char *type) = 0;
 
   // Instancing ///////////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIGroup newGroup() = 0;
+  virtual ANARIGroup newGroup() = 0;
 
-  ANARI_INTERFACE virtual ANARIInstance newInstance() = 0;
+  virtual ANARIInstance newInstance() = 0;
 
   // Top-level Worlds /////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIWorld newWorld() = 0;
+  virtual ANARIWorld newWorld() = 0;
 
   // Object + Parameter Lifetime Management ///////////////////////////////////
 
-  ANARI_INTERFACE virtual void setParameter(ANARIObject object,
+  virtual void setParameter(ANARIObject object,
       const char *name,
       ANARIDataType type,
       const void *mem) = 0;
 
-  ANARI_INTERFACE virtual void unsetParameter(
-      ANARIObject object, const char *name) = 0;
+  virtual void unsetParameter(ANARIObject object, const char *name) = 0;
 
-  ANARI_INTERFACE virtual void commit(ANARIObject object) = 0;
+  virtual void commit(ANARIObject object) = 0;
 
-  ANARI_INTERFACE virtual void release(ANARIObject _obj) = 0;
-  ANARI_INTERFACE virtual void retain(ANARIObject _obj) = 0;
+  virtual void release(ANARIObject _obj) = 0;
+  virtual void retain(ANARIObject _obj) = 0;
 
   // Object Query Interface ///////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual int getProperty(ANARIObject object,
+  virtual int getProperty(ANARIObject object,
       const char *name,
       ANARIDataType type,
       void *mem,
@@ -138,47 +135,44 @@ struct Device
 
   // FrameBuffer Manipulation /////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIFrame newFrame() = 0;
+  virtual ANARIFrame newFrame() = 0;
 
-  ANARI_INTERFACE virtual const void *frameBufferMap(
-      ANARIFrame fb, const char *channel) = 0;
+  virtual const void *frameBufferMap(ANARIFrame fb, const char *channel) = 0;
 
-  ANARI_INTERFACE virtual void frameBufferUnmap(
-      ANARIFrame fb, const char *channel) = 0;
+  virtual void frameBufferUnmap(ANARIFrame fb, const char *channel) = 0;
 
   // Frame Rendering //////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIRenderer newRenderer(const char *type) = 0;
+  virtual ANARIRenderer newRenderer(const char *type) = 0;
 
-  ANARI_INTERFACE virtual void renderFrame(ANARIFrame) = 0;
-  ANARI_INTERFACE virtual int frameReady(ANARIFrame, ANARIWaitMask) = 0;
-  ANARI_INTERFACE virtual void discardFrame(ANARIFrame) = 0;
+  virtual void renderFrame(ANARIFrame) = 0;
+  virtual int frameReady(ANARIFrame, ANARIWaitMask) = 0;
+  virtual void discardFrame(ANARIFrame) = 0;
 
   /////////////////////////////////////////////////////////////////////////////
   // Extension interface
   /////////////////////////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE virtual ANARIObject newObject(
-      const char *objectType, const char *type);
+  virtual ANARIObject newObject(const char *objectType, const char *type);
 
-  ANARI_INTERFACE virtual void (*getProcAddress(const char *name))(void);
+  virtual void (*getProcAddress(const char *name))(void);
 
   /////////////////////////////////////////////////////////////////////////////
   // Helper/other functions and data members
   /////////////////////////////////////////////////////////////////////////////
 
-  ANARI_INTERFACE Device() = default;
-  ANARI_INTERFACE virtual ~Device() = default;
+  Device() = default;
+  virtual ~Device() = default;
 
-  ANARI_INTERFACE static Device *createDevice(const char *type,
+  static Device *createDevice(const char *type,
       ANARIStatusCallback defaultStatusCB = nullptr,
       void *defaultStatusCBUserPtr = nullptr);
 
  protected:
-  ANARI_INTERFACE ANARIDevice this_device() const;
+  ANARIDevice this_device() const;
 
-  ANARI_INTERFACE ANARIStatusCallback defaultStatusCallback() const;
-  ANARI_INTERFACE void *defaultStatusCallbackUserPtr() const;
+  ANARIStatusCallback defaultStatusCallback() const;
+  void *defaultStatusCallbackUserPtr() const;
 
  public:
   // NOTE: Unsuccessful to get the declaration of anariNewDevice() declared

--- a/libs/anari/include/anari/detail/Library.h
+++ b/libs/anari/include/anari/detail/Library.h
@@ -15,7 +15,7 @@ void *loadANARILibrary(const std::string &libName);
 void *getSymbolAddress(void *lib, const std::string &symbol);
 void freeLibrary(void *lib);
 
-struct Library
+struct ANARI_INTERFACE Library
 {
   Library(
       const char *name, ANARIStatusCallback defaultStatusCB, void *statusCBPtr);
@@ -91,48 +91,44 @@ struct Library
 
 // [REQUIRED] Define the function which allocates Device objects by subtype
 #define ANARI_DEFINE_LIBRARY_NEW_DEVICE(libname, type)                         \
-  extern "C" ANARIDevice anari_library_##libname##_new_device(const char *type)
+  ANARIDevice anari_library_##libname##_new_device(const char *type)
 
 // [OPTIONAL] Define the initialization function when library is loaded.
-#define ANARI_DEFINE_LIBRARY_INIT(libname)                                     \
-  extern "C" ANARI_INTERFACE void anari_library_##libname##_init()
+#define ANARI_DEFINE_LIBRARY_INIT(libname) void anari_library_##libname##_init()
 
 // [OPTIONAL] Define the function which allocates library-specific storage on
 //            construction (after init).
 #define ANARI_DEFINE_LIBRARY_ALLOCATE(libname)                                 \
-  extern "C" ANARI_INTERFACE void *anari_library_##libname##_allocate()
+  void *anari_library_##libname##_allocate()
 
 // [OPTIONAL] Define the function which frees library-specific storage on
 //            destruction.
 #define ANARI_DEFINE_LIBRARY_FREE(libname, libdata)                            \
-  extern "C" ANARI_INTERFACE void anari_library_##libname##_free(void *libdata)
+  void anari_library_##libname##_free(void *libdata)
 
 // [OPTIONAL] Define the function which loads a library-specific module.
 #define ANARI_DEFINE_LIBRARY_LOAD_MODULE(libname, libdata, modname)            \
-  extern "C" ANARI_INTERFACE void anari_library_##libname##_load_module(       \
-      void *libdata, const char *modname)
+  void anari_library_##libname##_load_module(void *libdata, const char *modname)
 
 // [OPTIONAL] Define the function which unloads a library-specific module.
 #define ANARI_DEFINE_LIBRARY_UNLOAD_MODULE(libname, libdata, modname)          \
-  extern "C" ANARI_INTERFACE void anari_library_##libname##_unload_module(     \
+  void anari_library_##libname##_unload_module(                                \
       void *libdata, const char *modname)
 
 // [OPTIONAL] Define introspection functions.
 #define ANARI_DEFINE_LIBRARY_GET_DEVICE_SUBTYPES(libname, libdata)             \
-  extern "C" ANARI_INTERFACE const char *                                      \
-      *anari_library_##libname##_get_device_subtypes(void *libdata)
+  const char **anari_library_##libname##_get_device_subtypes(void *libdata)
 #define ANARI_DEFINE_LIBRARY_GET_OBJECT_SUBTYPES(                              \
     libname, libdata, deviceSubtype, objectType)                               \
-  extern "C" ANARI_INTERFACE const char *                                      \
-      *anari_library_##libname##_get_object_subtypes(                          \
-          void *libdata, const char *deviceSubtype, ANARIDataType objectType)
+  const char **anari_library_##libname##_get_object_subtypes(                  \
+      void *libdata, const char *deviceSubtype, ANARIDataType objectType)
 #define ANARI_DEFINE_LIBRARY_GET_OBJECT_PARAMETERS(                            \
     libname, libdata, deviceSubtype, objectSubtype, objectType)                \
-  extern "C" ANARI_INTERFACE const ANARIParameter                              \
-      *anari_library_##libname##_get_object_parameters(void *libdata,          \
-          const char *deviceSubtype,                                           \
-          const char *objectSubtype,                                           \
-          ANARIDataType objectType)
+  const ANARIParameter *anari_library_##libname##_get_object_parameters(       \
+      void *libdata,                                                           \
+      const char *deviceSubtype,                                               \
+      const char *objectSubtype,                                               \
+      ANARIDataType objectType)
 #define ANARI_DEFINE_LIBRARY_GET_PARAMETER_PROPERTY(libname,                   \
     libdata,                                                                   \
     deviceSubtype,                                                             \
@@ -142,15 +138,14 @@ struct Library
     parameterType,                                                             \
     propertyName,                                                              \
     propertyType)                                                              \
-  extern "C" ANARI_INTERFACE const void                                        \
-      *anari_library_##libname##_get_parameter_property(void *libdata,         \
-          const char *deviceSubtype,                                           \
-          const char *objectSubtype,                                           \
-          ANARIDataType objectType,                                            \
-          const char *parameterName,                                           \
-          ANARIDataType parameterType,                                         \
-          const char *propertyName,                                            \
-          ANARIDataType propertyType)
+  const void *anari_library_##libname##_get_parameter_property(void *libdata,  \
+      const char *deviceSubtype,                                               \
+      const char *objectSubtype,                                               \
+      ANARIDataType objectType,                                                \
+      const char *parameterName,                                               \
+      ANARIDataType parameterType,                                             \
+      const char *propertyName,                                                \
+      ANARIDataType propertyType)
 
 ANARI_TYPEFOR_SPECIALIZATION(Library *, ANARI_LIBRARY);
 

--- a/libs/anari/include/anari/detail/Library.h
+++ b/libs/anari/include/anari/detail/Library.h
@@ -95,39 +95,40 @@ struct Library
 
 // [OPTIONAL] Define the initialization function when library is loaded.
 #define ANARI_DEFINE_LIBRARY_INIT(libname)                                     \
-  extern "C" void anari_library_##libname##_init()
+  extern "C" ANARI_INTERFACE void anari_library_##libname##_init()
 
 // [OPTIONAL] Define the function which allocates library-specific storage on
 //            construction (after init).
 #define ANARI_DEFINE_LIBRARY_ALLOCATE(libname)                                 \
-  extern "C" void *anari_library_##libname##_allocate()
+  extern "C" ANARI_INTERFACE void *anari_library_##libname##_allocate()
 
 // [OPTIONAL] Define the function which frees library-specific storage on
 //            destruction.
 #define ANARI_DEFINE_LIBRARY_FREE(libname, libdata)                            \
-  extern "C" void anari_library_##libname##_free(void *libdata)
+  extern "C" ANARI_INTERFACE void anari_library_##libname##_free(void *libdata)
 
 // [OPTIONAL] Define the function which loads a library-specific module.
 #define ANARI_DEFINE_LIBRARY_LOAD_MODULE(libname, libdata, modname)            \
-  extern "C" void anari_library_##libname##_load_module(                       \
+  extern "C" ANARI_INTERFACE void anari_library_##libname##_load_module(       \
       void *libdata, const char *modname)
 
 // [OPTIONAL] Define the function which unloads a library-specific module.
 #define ANARI_DEFINE_LIBRARY_UNLOAD_MODULE(libname, libdata, modname)          \
-  extern "C" void anari_library_##libname##_unload_module(                     \
+  extern "C" ANARI_INTERFACE void anari_library_##libname##_unload_module(     \
       void *libdata, const char *modname)
 
 // [OPTIONAL] Define introspection functions.
 #define ANARI_DEFINE_LIBRARY_GET_DEVICE_SUBTYPES(libname, libdata)             \
-  extern "C" const char **anari_library_##libname##_get_device_subtypes(       \
-      void *libdata)
+  extern "C" ANARI_INTERFACE const char *                                      \
+      *anari_library_##libname##_get_device_subtypes(void *libdata)
 #define ANARI_DEFINE_LIBRARY_GET_OBJECT_SUBTYPES(                              \
     libname, libdata, deviceSubtype, objectType)                               \
-  extern "C" const char **anari_library_##libname##_get_object_subtypes(       \
-      void *libdata, const char *deviceSubtype, ANARIDataType objectType)
+  extern "C" ANARI_INTERFACE const char *                                      \
+      *anari_library_##libname##_get_object_subtypes(                          \
+          void *libdata, const char *deviceSubtype, ANARIDataType objectType)
 #define ANARI_DEFINE_LIBRARY_GET_OBJECT_PARAMETERS(                            \
     libname, libdata, deviceSubtype, objectSubtype, objectType)                \
-  extern "C" const ANARIParameter                                              \
+  extern "C" ANARI_INTERFACE const ANARIParameter                              \
       *anari_library_##libname##_get_object_parameters(void *libdata,          \
           const char *deviceSubtype,                                           \
           const char *objectSubtype,                                           \
@@ -141,15 +142,15 @@ struct Library
     parameterType,                                                             \
     propertyName,                                                              \
     propertyType)                                                              \
-  extern "C" const void *anari_library_##libname##_get_parameter_property(     \
-      void *libdata,                                                           \
-      const char *deviceSubtype,                                               \
-      const char *objectSubtype,                                               \
-      ANARIDataType objectType,                                                \
-      const char *parameterName,                                               \
-      ANARIDataType parameterType,                                             \
-      const char *propertyName,                                                \
-      ANARIDataType propertyType)
+  extern "C" ANARI_INTERFACE const void                                        \
+      *anari_library_##libname##_get_parameter_property(void *libdata,         \
+          const char *deviceSubtype,                                           \
+          const char *objectSubtype,                                           \
+          ANARIDataType objectType,                                            \
+          const char *parameterName,                                           \
+          ANARIDataType parameterType,                                         \
+          const char *propertyName,                                            \
+          ANARIDataType propertyType)
 
 ANARI_TYPEFOR_SPECIALIZATION(Library *, ANARI_LIBRARY);
 

--- a/libs/anari_utilities/CMakeLists.txt
+++ b/libs/anari_utilities/CMakeLists.txt
@@ -6,6 +6,7 @@ project(anari_utilities VERSION ${PROJECT_VERSION})
 add_library(${PROJECT_NAME}
   demangle.cpp
   ParameterizedObject.cpp
+  IntrusivePtr.cpp
 )
 
 target_link_libraries(${PROJECT_NAME} PUBLIC anari)
@@ -15,6 +16,8 @@ PUBLIC
   $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>
   $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
 )
+
+target_compile_definitions(${PROJECT_NAME} PRIVATE "anari_EXPORTS")
 
 ## Install library + targets
 

--- a/libs/anari_utilities/IntrusivePtr.cpp
+++ b/libs/anari_utilities/IntrusivePtr.cpp
@@ -1,0 +1,4 @@
+// Copyright 2021 The Khronos Group
+// SPDX-License-Identifier: Apache-2.0
+
+#include "anari/detail/IntrusivePtr.h"

--- a/libs/anari_utilities/include/anari/detail/IntrusivePtr.h
+++ b/libs/anari_utilities/include/anari/detail/IntrusivePtr.h
@@ -19,24 +19,24 @@ enum RefType
   ALL
 };
 
-class RefCounted
+class ANARI_INTERFACE RefCounted
 {
  public:
-  ANARI_INTERFACE RefCounted() = default;
-  ANARI_INTERFACE virtual ~RefCounted() = default;
+  RefCounted() = default;
+  virtual ~RefCounted() = default;
 
-  ANARI_INTERFACE RefCounted(const RefCounted &) = delete;
-  ANARI_INTERFACE RefCounted(RefCounted &&) = delete;
+  RefCounted(const RefCounted &) = delete;
+  RefCounted(RefCounted &&) = delete;
 
-  ANARI_INTERFACE RefCounted &operator=(const RefCounted &) = delete;
-  ANARI_INTERFACE RefCounted &operator=(RefCounted &&) = delete;
+  RefCounted &operator=(const RefCounted &) = delete;
+  RefCounted &operator=(RefCounted &&) = delete;
 
-  ANARI_INTERFACE void refInc(RefType = PUBLIC) const;
-  ANARI_INTERFACE void refDec(RefType = PUBLIC) const;
-  ANARI_INTERFACE uint64_t useCount(RefType = ALL) const;
+  void refInc(RefType = PUBLIC) const;
+  void refDec(RefType = PUBLIC) const;
+  uint64_t useCount(RefType = ALL) const;
 
  private:
-  ANARI_INTERFACE uint64_t internalRefCount() const;
+  uint64_t internalRefCount() const;
 
   static constexpr uint64_t m_publicRefMask = 0x00000000FFFFFFFF;
   static constexpr uint64_t m_internalRefMask = 0xFFFFFFFF00000000;

--- a/libs/anari_utilities/include/anari/detail/ParameterizedObject.h
+++ b/libs/anari_utilities/include/anari/detail/ParameterizedObject.h
@@ -16,18 +16,18 @@ constexpr int PARAM_LOCAL_STORAGE_SIZE = sizeof(float) * 3 * 4;
 
 struct ParameterizedObject
 {
-  ParameterizedObject() = default;
-  virtual ~ParameterizedObject() = default;
+  ANARI_INTERFACE ParameterizedObject() = default;
+  ANARI_INTERFACE virtual ~ParameterizedObject() = default;
 
   struct Param
   {
-    Param(const std::string &name);
-    ~Param() = default;
+    ANARI_INTERFACE Param(const std::string &name);
+    ANARI_INTERFACE virtual ~Param() = default;
 
     template <typename T>
     void set(const T &v);
 
-    void setDirect(Any v);
+    ANARI_INTERFACE void setDirect(Any v);
 
     std::string name;
     ANARIDataType type{ANARI_UNKNOWN};
@@ -38,23 +38,24 @@ struct ParameterizedObject
     bool locallyStored{false};
   };
 
-  bool hasParam(const std::string &name);
+  ANARI_INTERFACE bool hasParam(const std::string &name);
 
   template <typename T>
   void setParam(const std::string &name, const T &t);
 
-  Any getParamDirect(const std::string &name);
-  void setParamDirect(const std::string &name, Any v);
+  ANARI_INTERFACE Any getParamDirect(const std::string &name);
+  ANARI_INTERFACE void setParamDirect(const std::string &name, Any v);
 
   template <typename T>
   T getParam(const std::string &name, T valIfNotFound);
 
-  void removeParam(const std::string &name);
+  ANARI_INTERFACE void removeParam(const std::string &name);
 
-  void resetAllParamQueryStatus();
+  ANARI_INTERFACE void resetAllParamQueryStatus();
 
  protected:
-  Param *findParam(const std::string &name, bool addIfNotExist = false);
+  ANARI_INTERFACE Param *findParam(
+      const std::string &name, bool addIfNotExist = false);
 
   std::vector<std::shared_ptr<Param>>::iterator params_begin();
   std::vector<std::shared_ptr<Param>>::iterator params_end();

--- a/libs/anari_utilities/include/anari/detail/ParameterizedObject.h
+++ b/libs/anari_utilities/include/anari/detail/ParameterizedObject.h
@@ -14,20 +14,20 @@ namespace anari {
 
 constexpr int PARAM_LOCAL_STORAGE_SIZE = sizeof(float) * 3 * 4;
 
-struct ParameterizedObject
+struct ANARI_INTERFACE ParameterizedObject
 {
-  ANARI_INTERFACE ParameterizedObject() = default;
-  ANARI_INTERFACE virtual ~ParameterizedObject() = default;
+  ParameterizedObject() = default;
+  virtual ~ParameterizedObject() = default;
 
-  struct Param
+  struct ANARI_INTERFACE Param
   {
-    ANARI_INTERFACE Param(const std::string &name);
-    ANARI_INTERFACE virtual ~Param() = default;
+    Param(const std::string &name);
+    virtual ~Param() = default;
 
     template <typename T>
     void set(const T &v);
 
-    ANARI_INTERFACE void setDirect(Any v);
+    void setDirect(Any v);
 
     std::string name;
     ANARIDataType type{ANARI_UNKNOWN};
@@ -38,24 +38,23 @@ struct ParameterizedObject
     bool locallyStored{false};
   };
 
-  ANARI_INTERFACE bool hasParam(const std::string &name);
+  bool hasParam(const std::string &name);
 
   template <typename T>
   void setParam(const std::string &name, const T &t);
 
-  ANARI_INTERFACE Any getParamDirect(const std::string &name);
-  ANARI_INTERFACE void setParamDirect(const std::string &name, Any v);
+  Any getParamDirect(const std::string &name);
+  void setParamDirect(const std::string &name, Any v);
 
   template <typename T>
   T getParam(const std::string &name, T valIfNotFound);
 
-  ANARI_INTERFACE void removeParam(const std::string &name);
+  void removeParam(const std::string &name);
 
-  ANARI_INTERFACE void resetAllParamQueryStatus();
+  void resetAllParamQueryStatus();
 
  protected:
-  ANARI_INTERFACE Param *findParam(
-      const std::string &name, bool addIfNotExist = false);
+  Param *findParam(const std::string &name, bool addIfNotExist = false);
 
   std::vector<std::shared_ptr<Param>>::iterator params_begin();
   std::vector<std::shared_ptr<Param>>::iterator params_end();

--- a/libs/anari_utilities/include/anari/detail/demangle.h
+++ b/libs/anari_utilities/include/anari/detail/demangle.h
@@ -7,9 +7,12 @@
 #include <string>
 #include <typeinfo>
 
+// anari
+#include "anari/anari.h"
+
 namespace anari {
 
-std::string demangle(const char *name);
+ANARI_INTERFACE std::string demangle(const char *name);
 
 template <class T>
 inline std::string nameOf()

